### ISSUE TITLE
Distribution key as varchar causing problems

### ISSUE THREAD

**Issue by hghammoud** [citusdata/citus#6543]
I am currently facing an issue with tenant_id being a varchar instead of integer. I simplified the example 👍 

This works
```
CREATE TABLE event (
  tenant_id int,
  id bigint,
  primary key (tenant_id, id)
);

CREATE TABLE page (
  tenant_id int,
  id int,
  primary key (tenant_id, id)
);

SELECT create_distributed_table('event', 'tenant_id');
SELECT create_distributed_table('page', 'tenant_id', colocate_with => 'event');
alter table page add constraint fk21 foreign key (tenant_id, id) references event;
```

this does NOT work

```
CREATE TABLE event (
  tenant_id varchar,
  id bigint,
  primary key (tenant_id, id)
);

CREATE TABLE page (
  tenant_id varchar,
  id int,
  primary key (tenant_id, id)
);

SELECT create_distributed_table('event', 'tenant_id');
SELECT create_distributed_table('page', 'tenant_id', colocate_with => 'event');
alter table page add constraint fk21 foreign key (tenant_id, id) references event;
```

I am getting the following error
```
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
CONTEXT:  SQL statement "SELECT fk."tenant_id", fk."id" FROM ONLY "public"."page" fk LEFT OUTER JOIN ONLY "public"."event" pk ON ( pk."tenant_id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."tenant_id"::pg_catalog.text AND pk."id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."id"::pg_catalog.text) WHERE pk."tenant_id" IS NULL AND (fk."tenant_id" IS NOT NULL AND fk."id" IS NOT NULL)"
while executing command on docker-citus_worker-1:5432
```

I am using the  docker-compose found on the root of the github (citus 11.1)

What am I missing?

thank you.

**Comment by TsinghuaLucky912** [comment]
Dear Hackers, I have investigated this issue, and now I will tell you the results of my investigation for your reference:

0. Environment: Ubuntu22+pg15.0+citus main
1. I think the root cause of the error is: there is no pg_operator tuple for varchar type in postgresql (if the two tables here use text type, there is no such problem, because pg_operator has the content of text=text and its oid =98)
2. Because of the reason in #1, the generate_operator_clause function converts the varchar type to the text type, which leads to the following error. If you execute this select command (manually replace the text type with the varchar type, the query can be successful)
3. If you have any new ideas, please let me know, and I will see if I can solve this problem. Thank you!

```
src/backend/utils/adt/ruleutils.c

void
generate_operator_clause(StringInfo buf,
						 const char *leftop, Oid leftoptype,
						 Oid opoid,
						 const char *rightop, Oid rightoptype)
{
    // here
}
```

![调查结果](https://user-images.githubusercontent.com/49380232/205845356-f011bfb8-1c3c-482c-ab1f-18f9b3282d25.png)

**Comment by marcocitus** [comment]
I think the issue is specifically with the cast from varchar to text in the join clause.

```
SELECT fk."tenant_id", fk."id" FROM ONLY "public"."page" fk LEFT OUTER JOIN ONLY "public"."event" pk ON ( pk."tenant_id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."tenant_id"::pg_catalog.text AND pk."id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."id"::pg_catalog.text) WHERE pk."tenant_id" IS NULL AND (fk."tenant_id" IS NOT NULL AND fk."id" IS NOT NULL)
ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
```

**Comment by TsinghuaLucky912** [comment]
> I think the issue is specifically with the cast from varchar to text in the join clause.
> 
> ```
> SELECT fk."tenant_id", fk."id" FROM ONLY "public"."page" fk LEFT OUTER JOIN ONLY "public"."event" pk ON ( pk."tenant_id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."tenant_id"::pg_catalog.text AND pk."id"::pg_catalog.text OPERATOR(pg_catalog.=) fk."id"::pg_catalog.text) WHERE pk."tenant_id" IS NULL AND (fk."tenant_id" IS NOT NULL AND fk."id" IS NOT NULL)
> ERROR:  complex joins are only supported when all distributed tables are co-located and joined on their distribution columns
> ```

Yes, if you still keep the varchar type, there is no problem

**Comment by TsinghuaLucky912** [comment]
@marcocitus My feeling is that the problem is not citus, but PostgreSQL's missing handling of varchar types

![1670408501677_4CA3B77F-D6BC-478c-A608-7E00DAF061B0](https://user-images.githubusercontent.com/49380232/206154178-2be20bda-975f-4af0-a5a2-faf4fc932974.png)

![lQLPJwGCYg6DfCzNA-3NB32wsBZA4SMCYtUDjKXKhMDqAA_1917_1005](https://user-images.githubusercontent.com/49380232/206154310-f9c1714c-9d15-4a3c-aa13-5107f7e8caca.png)

![lQLPJwZIvpRCK6zNA-3NB3mw0CV86y37L_sDjKXIdEBVAA_1913_1005](https://user-images.githubusercontent.com/49380232/206154343-37e7c475-137d-4208-b9b8-ff34c2203c24.png)

**Comment by JelteF** [comment]
I have a fix ready in #6550
